pkgs/by-name/sw/switch-to-configuration-ng/src/src/main.rs

```diff
@@ -1662,30 +1651,45 @@ won't take effect until you reboot the system.
                     &user_dbus_path,
                     Duration::from_millis(5000),
                 );
-                let gid: u32 = proxy
-                    .get("org.freedesktop.login1.User", "GID")
-                    .with_context(|| format!("Failed to get GID for {name}"))?;
 
                 let runtime_path: String = proxy
                     .get("org.freedesktop.login1.User", "RuntimePath")
                     .with_context(|| format!("Failed to get runtime directory for {name}"))?;
 
+                let dbus_socket_path = PathBuf::from_str(&runtime_path)
+                    .with_context(|| format!("Invalid user runtime path {}", runtime_path))?
+                    .join("bus");
+
+                let dbus_socket_path = format!(
+                    "unix:path={}",
+                    dbus_socket_path.to_str().with_context(|| format!(
+                        "Invalid UTF-8 in DBus socket path {}",
+                        dbus_socket_path.display()
+                    ))?
+                );
+
+                seteuid(uid.into()).with_context(|| format!("Failed to seteuid to {}", uid))?;
```

Comment:
Is this atomic? i.e. If it fails, are we certain to be back in the original euid? Or could we accidentally be continuing with the wrong euid? What are the consequences if so?

Reply:
As is right now, if `seteuid()` fails the entire program fails

Reply:
Which, to be fair, is inconsistent with the perl script. https://github.com/jmbaur/nixpkgs/blob/212e3ccc5367b5933335fa2c8e33e53e8cd03571/nixos/modules/system/activation/switch-to-configuration.pl#L892 does not fail hard if any bit of what's going on there fails, so we should probably do the same. Agreed?

Reply:
To reiterate what we discussed on matrix, the `?` operator is much more lenient than just exiting the program. We need to ensure that this is safe in terms of *how this function can fail*, as well as *what happens when it does*.

Reply:
(And I understand that maybe the answers to those questions are obvious, but I don't believe it's obvious to the nixos teams, so it ought to be proven somehow first).

Reply:
`nix`'s seteuid appears to do the right thing: https://docs.rs/nix/latest/src/nix/unistd.rs.html#1487-1491, any non-zero syscall return value should return a `Result::Err`. I've changed our code to be more consistent with the perl script. Any failure to do any operation relating to user switching doesn't fail the entire stc program, it will just continue to the next user.

Reply:
The program is strictly less dangerous when it lowers its effective UID, because as `root` it could do anything (except pass DBus policy checks, apparently), but as a specific user it cannot. Therefore I don’t really have any concerns about this.

I suppose we could make a guard type returned from a `seteuid` wrapper that puts the EUID back on destruction and rely on RAII? Or even just one that takes a callback to run as the user. That would be cleaner. But again I don’t see why there should be a security concern here; if DBus didn’t have weird policy, all this code would just run as `root` instead.

Reply:
I do believe that failing to switch back to EUID 0 should cause the program to panic though, just on the grounds that we don’t want to execute the rest of the program in a weird unknown state that isn’t going to work.

Reply:
Given that we want to have error handling around setting the euid back to 0, I don't think we should perform the seteuid(0) in a Drop impl.

Reply:
Yes, good point. The callback approach however would still be viable.

Reply:
> The program is strictly less dangerous when it lowers its effective UID

I don't think that's true when it restores root perms later

Reply:
We discussed this more on matrix. I'm still a tad uncertain, but not in a way that makes me want to block this PR. So it's all good to me.

Reply:
On the other hand, even though I think this is safe I’ve moved more in the direction of just having a separate binary to do this :) We could drop the `__NIXOS_SWITCH_TO_CONFIGURATION_PARENT_EXE` stuff, which is really the worst part of it.

Reply:
What was the concern with `__NIXOS_SWITCH_TO_CONFIGURATION_PARENT_EXE`? On a technical level I don't see a difference between creating a child process that calls /proc/self/exe that conditions it's behavior on its environment versus calling /some/other/program that has the same behavior.

Reply:
We wouldn’t have to have any kind of environment variable guard for incorrect usage or conditionalize a simple UID check on it. We can just call out to a program that does its specific job and nothing else. Which makes sense, since the two code‐paths are entirely disjoint. It would also avoid depending on `/proc`, although that’s… marginal.

Just taking a `--do-user-activation-no-really-trust-me` argument or similar rather than looking at the UID would accomplish the same with less bureaucracy than currently, too. But doing it this way would also help make the file less big, so it might be a good idea in general.

Reply:
Regardless of implementation, I think the child process would want to have some way of making it less likely of being discovered by users, whether it be a runtime check or hiding it in libexec. I don't think it's in our best interest to support running it when not called from stc.

Reply:
The package won’t be installed on end‐user systems anyway, right? It’s not in the `$PATH` or anything. So it’s just an opaque store path that… does something pretty harmless that would happen when you logged in anyway, if you happen to track it down. But `libexec` would be maximally correct, yeah.

Reply:
The behavior of it today is harmless to run on its own, but who knows in the future. Just don't want to open up any new interfaces that start getting used. It's probably bike shedding at that point though, so likely not a big deal :)